src/test/java/com/puppycrawl/tools/checkstyle/internal/ArchUnitTest.java

```diff
@@ -74,4 +120,36 @@ public void nonProtectedCheckMethodsTest() {
 
         checkMethodsShouldNotBeProtectedRule.check(importedClasses);
     }
+
+    /**
+     * The goal is to ensure all classes in api package are not dependent on classes in util
+     * packages.
+     */
+    @Test
+    public void testClassesInApiDoNotDependOnClassesInUtil() {
+        final JavaClasses apiPackage = new ClassFileImporter()
+            .withImportOption(ImportOption.Predefined.DO_NOT_INCLUDE_TESTS)
+            .importPackages("com.puppycrawl.tools.checkstyle.api");
+
+        final String[] utilPackages = {
+            "com.puppycrawl.tools.checkstyle.utils",
+            "com.puppycrawl.tools.checkstyle.checks.javadoc.utils",
+        };
+
+        final ArchRule classShouldNotDependOnUtilPackages = noClasses()
+            .should()
+            .dependOnClassesThat()
+            .resideInAnyPackage(utilPackages);
+
+        final EvaluationResult result = classShouldNotDependOnUtilPackages.evaluate(apiPackage);
+        final EvaluationResult filtered = result.filterDescriptionsMatching(description -> {
+            return API_PACKAGE_SUPPRESSION_DETAILS.stream()
+                .noneMatch(description::startsWith);
+        });
```

Comment:
also it would be good to print useless suppressions, so it forces us to remove them if any.
but it can be implemented later, no need to do it right now :)

Reply:
Let's implement it later, and do it in a separate issue where we would also make better formatting of errors if any.

Reply:
@Vyom-Yadav , please create issue on this to not forget.
One day we will find contributor to make it happen.

Reply:
@Vyom-Yadav did we create issue on this?

Reply:
Created at https://github.com/checkstyle/checkstyle/issues/12189